src/scss/parse.js

```diff
@@ -4038,106 +4023,257 @@ function _checkUnicodeWildcard(i) {
+    case TokenType.QuestionMark:
+    case TokenType.CommercialAt:
+    case TokenType.LeftSquareBracket:
+    case TokenType.RightSquareBracket:
+    case TokenType.LowLine:
+    case TokenType.Tilde:
+      return 1;
+
+    default:
+      return 0;
+  }
+}
+
+/**
+ * Check if content of URI can be contained within a raw node
+ * @param {number} i Token's index number
+ * @returns {number} Length of raw node
+ */
+function checkUriRaw(i) {
+  const start = i;
   let l;
-  let raw;
 
-  pos += 2;
+  while (i < tokensLength) {
+    if (checkInterpolation(i)) break;
 
-  uriExcluding[TokenType.Space] = 1;
-  uriExcluding[TokenType.Tab] = 1;
-  uriExcluding[TokenType.Newline] = 1;
-  uriExcluding[TokenType.LeftParenthesis] = 1;
-  uriExcluding[TokenType.RightParenthesis] = 1;
-
-  if (checkUriContent(pos)) {
-    uri = []
-        .concat(getSC())
-        .concat(getUriContent())
-        .concat(getSC());
-  } else {
-    uri = [].concat(getSC());
-    l = checkExcluding(uriExcluding, pos);
-    token = tokens[pos];
-    raw = newNode(NodeType.RawType, joinValues(pos, pos + l), token.ln,
-        token.col);
+    if (l = checkIdent(i)) i += l;
+    else if (l = checkNumber(i)) i += l;
+    else if (l = _checkUriRawCharacters(i)) i += l;
+    // As comments are not allowed within the uri token, include as part of node
+    else if (l = checkCommentML(i)) i += l;
+    else break;
+  }
 
-    uri.push(raw);
+  return i - start;
+}
 
-    pos += l + 1;
+/**
+ * Get a raw node
+ * @returns {Node}
+ */
+function getUriRaw() {
+  const startPos = pos;
+  const type = NodeType.RawType;
+  const token = tokens[startPos];
+  const line = token.ln;
+  const column = token.col;
+  let content = [];
+  let l;
+
+  while (pos < tokensLength) {
+    if (checkInterpolation(pos)) break;
 
-    uri = uri.concat(getSC());
+    if (checkIdent(pos)) getIdent();
+    else if (checkNumber(pos)) getNumber();
+    else if (_checkUriRawCharacters(pos)) pos += 1;
+    // As comments are not allowed within the uri token, include as part of node
+    else if (l = checkCommentML(pos)) pos += l;
+    else break;
   }
 
-  token = tokens[startPos];
-  var line = token.ln;
-  var column = token.col;
-  var end = getLastPosition(uri, line, column, 1);
-  pos++;
+  content = joinValues(startPos, pos - 1);
 
-  return newNode(NodeType.UriType, uri, token.ln, token.col, end);
+  return newNode(type, content, line, column);
 }
 
 /**
- * @param {Number} i Token's index number
- * @returns {Number}
+ * Check for a raw URI
+ * (1) `url(http://foo/bar.png)`
+ * (2) `url(http://foo/#{bar}.png)`
+ * (3) `url(#{foo}/bar.png)`
+ * @param {number} i Token's index number
+ * @returns {number} Length of URI node
  */
-function checkUriContent(i) {
-  return checkUri1(i) ||
-      checkFunction(i);
+function checkUri1(i) {
+  const start = i;
+  let l;
+  let isRaw = false;
+
+  if (l = checkSC(i)) i += l;
+
+  while (i < tokensLength) {
+    if (checkVariable(i) || checkFunction(i)) return 0;
```

Comment:
I think this line can be removed. We don't need to check for function because parentheses are invalid raw tokens. Plus we should not break on variables because `url($a)` is a perfectly valid piece of css code where `$a` is just part of uri and not a variable.